File: python/day4.py
def find_xmas(x: int, y: int, dx: int, dy: int, lines: list[list[str]]) -> bool:
    word = "XMAS"

    x_max, y_max = x + 3 * dx, y + 3 * dy
    if x_max >= len(lines[0]) or x_max < 0:
        return False
    if y_max >= len(lines) or y_max < 0:
        return False

    for i in range(4):
        if lines[y + i * dy][x + i * dx] != word[i]:
            return False

    return True


def find_mas(
    x: int, y: int, dx1: int, dy1: int, dx2: int, dy2: int, lines: list[list[str]]
) -> bool:
    if x + 1 >= len(lines[0]) or x - 1 < 0:
        return False
    if y + 1 >= len(lines) or y - 1 < 0:
        return False

    if lines[y][x] == "A":
        if lines[y + dy1][x + dx1] == "M" and lines[y + dy2][x + dx2] == "M":
            if lines[y - dy1][x - dx1] == "S" and lines[y - dy2][x - dx2] == "S":
                return True

    return False


def day4a() -> None:
    lines = []
    with open("../inputs/day4.txt", "r") as file:
        for line in file:
            lines.append(line.strip())

    dirs = [
        (0, 1),
        (0, -1),
        (1, 1),
        (1, 0),
        (1, -1),
        (-1, 1),
        (-1, 0),
        (-1, -1),
    ]

    num_xmas = 0
    for y in range(len(lines)):
        for x in range(len(lines[0])):
            for dir in dirs:
                if find_xmas(x, y, dir[0], dir[1], lines):
                    num_xmas += 1

    print(num_xmas)


def day4b() -> None:
    lines = []
    with open("../inputs/day4.txt", "r") as file:
        for line in file:
            lines.append(line.strip())

    dirs = [
        (1, 1, -1, 1),
        (-1, 1, -1, -1),
        (-1, -1, 1, -1),
        (1, -1, 1, 1),
    ]

    num_mas = 0
    for y in range(len(lines)):
        for x in range(len(lines[0])):
            for dir in dirs:
                if find_mas(x, y, dir[0], dir[1], dir[2], dir[3], lines):
                    num_mas += 1

    print(num_mas)


day4b()

File: go/day4.go
package main

import "fmt"

func checkXmas(x, y, dx, dy int, lines []string) bool {
	maxX, maxY := x+3*dx, y+3*dy
	if maxX < 0 || maxX >= len(lines[0]) {
		return false
	}
	if maxY < 0 || maxY >= len(lines) {
		return false
	}

	for i, char := range "XMAS" {
		if lines[y+i*dy][x+i*dx] != byte(char) {
			return false
		}
	}

	return true
}

func checkMas(x, y, dx1, dy1, dx2, dy2 int, lines []string) bool {
	if x-1 < 0 || x+1 >= len(lines[0]) {
		return false
	}
	if y-1 < 0 || y+1 >= len(lines) {
		return false
	}

	if lines[y][x] != 'A' {
		return false
	}
	if lines[y+dy1][x+dx1] != 'M' || lines[y+dy2][x+dx2] != 'M' {
		return false
	}
	if lines[y-dy1][x-dx1] != 'S' || lines[y-dy2][x-dx2] != 'S' {
		return false
	}

	return true
}

func Day4a() {
	lines, err := ReadLines("../inputs/day4.txt")
	if err != nil {
		panic("Could not read lines")
	}

	dirs := [][2]int{
		{0, 1},
		{0, -1},
		{1, 1},
		{1, 0},
		{1, -1},
		{-1, 1},
		{-1, 0},
		{-1, -1},
	}

	num_xmas := 0
	for i, line := range lines {
		for j := range line {
			for _, dir := range dirs {
				if checkXmas(j, i, dir[0], dir[1], lines) {
					num_xmas += 1
				}
			}
		}
	}

	fmt.Printf("Num XMAS: %v\n", num_xmas)
}

func Day4b() {
	lines, err := ReadLines("../inputs/day4.txt")
	if err != nil {
		panic("Could not read lines")
	}

	dirs := [][4]int{
		{1, -1, -1, -1},
		{-1, -1, -1, 1},
		{-1, 1, 1, 1},
		{1, 1, 1, -1},
	}

	num_xmas := 0
	for i, line := range lines {
		for j := range line {
			for _, dir := range dirs {
				if checkMas(j, i, dir[0], dir[1], dir[2], dir[3], lines) {
					num_xmas += 1
				}
			}
		}
	}

	fmt.Printf("Num X-MAS: %v\n", num_xmas)
}
